src/main/java/de/tum/in/www1/artemis/web/rest/TextClusterResource.java

```diff
@@ -0,0 +1,95 @@
+import de.tum.in.www1.artemis.service.AuthorizationCheckService;
+import de.tum.in.www1.artemis.web.rest.dto.TextClusterStatisticsDTO;
+
+/**
+ * REST controller for managing TextClusterResource.
+ */
+@RestController
+@RequestMapping("api/")
+public class TextClusterResource {
+
+    private final Logger log = LoggerFactory.getLogger(TextClusterResource.class);
+
+    private final TextClusterRepository textClusterRepository;
+
+    private final UserRepository userRepository;
+
+    private final ExerciseRepository exerciseRepository;
+
+    private final AuthorizationCheckService authCheckService;
+
+    public TextClusterResource(TextClusterRepository textClusterRepository, UserRepository userRepository, ExerciseRepository exerciseRepository,
+            AuthorizationCheckService authCheckService) {
+        this.textClusterRepository = textClusterRepository;
+        this.userRepository = userRepository;
+        this.exerciseRepository = exerciseRepository;
+        this.authCheckService = authCheckService;
+    }
+
+    /**
+     * Get text-exercises/{exerciseId}/cluster-statistics/
+     * <p>
+     * Get text cluster stats
+     * @param exerciseId The id of the text exercise to fetch cluster statistics data from
+     * @return The list of cluster ids adjacent to their respective sizes and automatically graded text blocks
+     */
+    @GetMapping("text-exercises/{exerciseId}/cluster-statistics")
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    public ResponseEntity<List<TextClusterStatisticsDTO>> getClusterStats(@PathVariable Long exerciseId) {
+        // Check if instructor has permission to access the exercise with given exerciseId
+        Exercise exercise = exerciseRepository.findByIdElseThrow(exerciseId);
+        User user = userRepository.getUserWithGroupsAndAuthorities();
+        authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, exercise, user);
+
+        // Get cluster statistics without the disabled state
+        List<TextClusterStatisticsDTO> clusterStats = textClusterRepository.getClusterStatistics(exerciseId);
+
+        // Fetch cluster with id and disabled state
+        Map<Long, Boolean> clusterIdAndDisabled = textClusterRepository.getTextClusterWithIdAndDisabled();
+
+        // Set cluster statistics disabled state
+        clusterStats.forEach(s -> s.setDisabled(clusterIdAndDisabled.get(s.getClusterId())));
+        log.debug("REST request to get clusterStats : {}", clusterStats);
+        return ResponseEntity.ok().body(clusterStats);
+    }
+
+    /**
+     * Patch text-clusters/{clusterId}
+     * Sets a text cluster's disabled boolean value
+     *
+     * @param clusterId The id of the cluster to be disabled/enabled
+     * @param disabled The predicate value defining the disabled state of the cluster
+     * @return The status whether the boolean value was set successfully or the setting failed.
+     */
+    @PatchMapping("text-clusters/{clusterId}")
```

Comment:
please add `text-exercises/{exerciseId}` in the beginning and use the exerciseId for the security check. Then also make sure the cluster's text exercise id is the same as exerciseId

Reply:
Addded the check [here](https://github.com/ls1intum/Artemis/pull/3910/commits/dabea957876752555a7c691218c5a6a38c5ebde7)